odoo/cli/module.py

```diff
@@ -0,0 +1,215 @@
+            case 'upgrade':
+                self._upgrade(parsed_args)
+            case 'force-demo':
+                self._force_demo(parsed_args)
+
+    def _get_zip_path(self, path):
+        if (
+            (fullpath := Path(path).resolve())
+            and fullpath.exists()
+            and fullpath.suffix.lower() == '.zip'
+        ):
+            return fullpath
+        return None
+
+    def _get_module_names(self, module_names):
+        """ Get valid module names from disk before starting the Db environment """
+        initialize_sys_path()
+        module_names = set(module_names)
+        if not module_names:
+            return module_names
+        valid_module_names = {
+            module
+            for module in module_names
+            if get_module_path(module)
+            or self._get_zip_path(module)
+        }
+        if not valid_module_names:
+            _logger.warning("No valid module names found")
+        return valid_module_names
+
+    def _get_modules(self, env, module_names, only_outdated=False):
+        """ Get modules from the db.
+
+            Note: the two version fields' names are inverted
+            as stated in a comment on ir.module.module.
+
+                Roses are red and violets are blue.
+                ``latest_version`` is ``installed_version``, true.
+                That's how it goes sometimes, in Odoo.
+                Is it a bug or a feature, it's up to you.
+
+            :param module_names:  name of the modules
+                                  the upgrade command has a default value '*'
+                                  to indicate all installed modules
+            :param only_outdated: filter only modules that have a newer version on disk.
+        """
+        Module = env['ir.module.module']
+        Module.update_list()
+        domain = []
+        if module_names == '*':
+            # `modules` is optional for upgrade
+            domain.append(('state', '=', 'installed'))
+        elif module_names:
+            domain.append(('name', 'in', module_names))
+        else:
+            return Module
+        modules = Module.search(domain)
+        if only_outdated:
+            modules = modules.filtered(
+                lambda x: parse_version(x.installed_version)
+                    > parse_version(x.latest_version),
+            )
+        return modules
+
+    @contextmanager
+    def _create_env_context(self, db_name):
+        with Registry.new(db_name).cursor() as cr:
+            yield Environment(cr, SUPERUSER_ID, {})
+
+    def _install(self, parsed_args):
+        with self._create_env_context(parsed_args.db_name) as env:
+
+            installable_modules = self._get_modules(env, parsed_args.modules)
+            parsed_args.modules = set(parsed_args.modules) - set(installable_modules.mapped("name"))
+            if installable_modules:
+                installable_modules.button_immediate_install()
+
+            importable_zipfiles = {
```

Comment:
use orderedset
since the imported zip file may need special installing order to meet their dependency